src/bsoncxx/test/bson_get_values.cpp

```diff
@@ -0,0 +1,154 @@
+#include "catch.hpp"
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/builder/stream/helpers.hpp>
+
+using namespace bsoncxx;
+
+TEST_CASE("[] can reach into nested arrays", "[bsoncxx]") {
+    using namespace builder::stream;
+
+    builder::stream::document build_doc;
+    // {
+    //     "ints": [ 1, 3, [ 5 ] ],
+    //     "bools": [ true, false ]
+    // }
+    build_doc << "ints" << open_array << 1 << 3 << open_array << 5 << close_array << close_array
+              << "bools" << open_array << true << false << close_array;
+
+    auto doc = build_doc.view();
+
+    SECTION("works with one level") {
+        REQUIRE(doc["ints"][0]);
+        REQUIRE(doc["ints"][0].get_int32() == 1);
+
+        REQUIRE(doc["ints"][1]);
+        REQUIRE(doc["ints"][1].get_int32() == 3);
+
+        REQUIRE(doc["bools"][0]);
+        REQUIRE(doc["bools"][0].get_bool() == true);
+
+        REQUIRE(doc["bools"][1]);
+        REQUIRE(doc["bools"][1].get_bool() == false);
+    }
+
+    SECTION("works with two levels") {
+        REQUIRE(doc["ints"][2][0]);
+        REQUIRE(doc["ints"][2][0].get_int32() == 5);
+    }
+
+    SECTION("throws on non-document") {
+        REQUIRE_THROWS(doc["ints"][0][0]);
+        REQUIRE_THROWS(doc["ints"][2][0][0]);
+        REQUIRE_THROWS(doc["bools"][1][0]);
+    }
+
+    SECTION("returns invalid on not found") {
+        REQUIRE(!doc["ints"][9]);
+        REQUIRE(!doc["ints"][2][9]);
+        REQUIRE(!doc["bools"][9]);
+    }
+}
+
+TEST_CASE("[] can reach into nested documents", "[bsoncxx]") {
```

Comment:
It is very important, while working in the bson library, to be aware of the distinction between BSON the spec, and BSON the document format used by MongoDB. People may want to use our BSON library in contexts that have nothing to do with MongoDB. The BSON library should implement the BSON spec, and be agnostic about database limits. If such limits are to be enforced, they should adhere at mongocxx level, not bsoncxx.